src/poetry/config/config.py

```diff
@@ -227,7 +227,7 @@ def get(self, setting_name: str, default: Any = None) -> Any:
                 if repositories:
                     return repositories
 
-            env = "POETRY_" + "_".join(k.upper().replace("-", "_") for k in keys)
+            env = "POETRY_" + "_".join(re.sub(r"[\W]", "_", k.upper()) for k in keys)
```

Comment:
per https://github.com/python-poetry/poetry/issues/7123#issuecomment-1332587118, this doesn't do anything for people who retrieve their configuration through the configuration file rather than through environment variables.  

Eg dots remain problematic, per the code a dozen lines up